### ISSUE TITLE
ONNX with ML.Net very slow performance in GPU

### ISSUE THREAD

**Issue by sportbilly21** [dotnet/machinelearning#6111]
System information
Windos 10 build 19044
ML.Net 1.7.0
Microsoft.ML.OnnxRuntime.GPU 1.10.0

 Issue

Currently our application is in .Net and for deploying the models and getting the predictions we are using Flask and python. The models this way are running very fast all models run in around 10ms and GPU utilizsation is very low. GPU is a Quadro P400. We are trying to remove Flask and run everything within .NET so I have converted 4 models trained in Tensorflow/Keras to ONNX and I using them for inference with the use of ML.Net. I have created the pipeline and the prediction engine. I have installed CUDA and CuDNN needed for the package. I have both 10.2 and 11 Cuda versions in the PC. Performance when running with ML.Net is very poor and GPU is utilisation are 50%. Any suggestions or thoughts regarding this

Code of pipeline and prediction engine:
            

                var pipeline = this.mlContext.Transforms.ResizeImages("input", height, width, inputColumnName: nameof(ImageData.ImgInput), resizing: ImageResizingEstimator.ResizingKind.Fill)// inputColumnName: nameof(ImageInputData.Image))
                .Append(mlContext.Transforms.ExtractPixels(inputLayer, "input", colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARGB, interleavePixelColors: true, outputAsFloatArray: true))
                    .Append(mlContext.Transforms.ApplyOnnxModel(outputLayer, inputLayer, modelFile, gpuDeviceId: gpuID));
                //var pipeline = this.mlContext.Transforms.ExtractPixels(inputLayer, inputColumnName: nameof(ImageData.ImgInput), colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARBG, interleavePixelColors: true);

                var data = mlContext.Data.LoadFromEnumerable(new List<ImageData>());
                var transformedData = pipeline.Fit(data).Transform(data);
                //var data = mlContext.Data.LoadFromEnumerable(new ImageData());
                this.model = pipeline.Fit(data); //;
                this.categories = cats;
                this.predictor = this.mlContext.Model.CreatePredictionEngine<ImageData, ImagePrediction>(this.model);

**Comment by michaelgsharp** [comment]
Can you share a full repro so we can take a look?

**Comment by sportbilly21** [comment]
I cannot share any images or models for this. I can share the file the file I am building the pipeline  
I using 4 onnx models so I am building 4 predictions engines (4 instances of this class)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text;
using Microsoft.ML;
using Microsoft.ML.Data;
using Microsoft.ML.Transforms.Image;

namespace ImgProc.DL
{
    class OnnxClassifier
    {

        private const int imageHeight = 224;
        private const int imageWidth = 224;
        private const int numberOfChannels = 3;
        public int inputSize = 0;
        private const int inputS = 4985;
        string modelFile;
        
        private static float[] predictions;

        public float[] GetPredictions
        {
            get { return predictions; }
        }

        private MLContext mlContext = new MLContext();

        private ITransformer model;

        public string columnName;

        private List<string> categories = new List<string>();

        private IDataView transformedValues;

        private PredictionEngine<ImageData, ImagePrediction> predictor;



        public List<string> GetCategoryNames
        {
            get { return categories; }
        }


        public void GenerateModel( string modelLocation, string inputLayer, string outputLayer, int height, int width, bool grey, List<string> cats=null, int gpuID = 0)
        {
            this.modelFile = Path.Combine(Directory.GetParent(System.IO.Directory.GetCurrentDirectory()).Parent.Parent.Parent.FullName, modelLocation);
            
#
            //LoadRawImageBytes(outputColumnName: "input", imageFolder: _imagesFolder, inputColumnName: nameof(ImageInputData.Image))
            if (!grey)
            {
                //var pipeline = this.mlContext.Transforms.ConvertToImage(height, width, outputColumnName: "input", inputColumnName: nameof(ImageData.ImgInput), interleavedColors: false)
                //.Append(mlContext.Transforms.ResizeImages(inputLayer,  height, width, resizing: ImageResizingEstimator.ResizingKind.IsoCrop, inputColumnName: "input"))// inputColumnName: nameof(ImageInputData.Image))
                //.Append(mlContext.Transforms.ExtractPixels(inputLayer, "input", colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARBG, interleavePixelColors: true, outputAsFloatArray: false))
                //.Append(mlContext.Transforms.ApplyOnnxModel(outputLayer, inputLayer, modelFile, gpuDeviceId: gpuID));

                var pipeline = this.mlContext.Transforms.ResizeImages("input", height, width, inputColumnName: nameof(ImageData.ImgInput), resizing: ImageResizingEstimator.ResizingKind.Fill)// inputColumnName: nameof(ImageInputData.Image))
                .Append(mlContext.Transforms.ExtractPixels(inputLayer, "input", colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARGB, interleavePixelColors: true, outputAsFloatArray: true))
                    .Append(mlContext.Transforms.ApplyOnnxModel(outputLayer, inputLayer, modelFile, gpuDeviceId: gpuID));
                //var pipeline = this.mlContext.Transforms.ExtractPixels(inputLayer, inputColumnName: nameof(ImageData.ImgInput), colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARBG, interleavePixelColors: true);

                var data = mlContext.Data.LoadFromEnumerable(new List<ImageData>());
                var transformedData = pipeline.Fit(data).Transform(data);
                //var data = mlContext.Data.LoadFromEnumerable(new ImageData());
                this.model = pipeline.Fit(data); //;
                this.categories = cats;
                this.predictor = this.mlContext.Model.CreatePredictionEngine<ImageData, ImagePrediction>(this.model);

            }
            if (grey)
            {
               var pipeline = this.mlContext.Transforms.ResizeImages("input", height, width, resizing: ImageResizingEstimator.ResizingKind.IsoCrop, inputColumnName: nameof(ImageData.ImgInput))// inputColumnName: nameof(ImageInputData.Image)) resizing: ImageResizingEstimator.ResizingKind.IsoPad, cropAnchor: ImageResizingEstimator.Anchor.Center
                .Append(mlContext.Transforms.ConvertToGrayscale("input", "input"))
                .Append(mlContext.Transforms.ExtractPixels(inputLayer, "input", colorsToExtract: ImagePixelExtractingEstimator.ColorBits.Rgb, orderOfExtraction: ImagePixelExtractingEstimator.ColorsOrder.ARGB, interleavePixelColors: true, offsetImage:117))
                .Append(mlContext.Transforms.ApplyOnnxModel(outputLayer, inputLayer, modelFile, gpuDeviceId: gpuID));

                var data = mlContext.Data.LoadFromEnumerable(new List<ImageData>());

                this.model = pipeline.Fit(data);
                this.categories = cats;
                this.predictor = this.mlContext.Model.CreatePredictionEngine<ImageData, ImagePrediction>(this.model);
            }

        }

        


        public void ClassifySingleImage(Bitmap imgClassify)
        {
            // load the fully qualified image file name into ImageData

            var imageData = new ImageData()
            {
                ImgInput = imgClassify
            };
            //var predictor = this.mlContext.Model.CreatePredictionEngine<ImageData, ImagePrediction>(this.model);
            var prediction = this.predictor.Predict(imageData);
            predictions = prediction.Preds;

        }


        public class ImageData
        {
            //[VectorType(224*224*3)]
            [ImageType(224, 224)]
            public Bitmap ImgInput { get; set; }


            [LoadColumn(1)]
            public string Label { get; set; }
        }

    }

    public class ImageInputData
    {

        public Bitmap Image { get; set; }
    }



    class ImagePrediction
    {
       //IDataView column = new IDataView()

        //Need to check how we can pass the column from the sortmap file 
        //[ColumnName("dense")]
        [ColumnName("dense")]
        public float[] Preds;
    }

 
}

**Comment by sportbilly21** [comment]
Hi, I tried using a Tensoflow model instead of ONNX but I am getting the same issue with the performance. @michaelgsharp does this issue has anything to do with you comment in #10142 in microsoft/onnxrutime git issue?
"Yeah, I'm one of the devs on the ML.NET side of things. This is actually an issue on our side. After you save/load a model, it no longer uses the GPU and the user doesn't have a way of forcing it to use the GPU. We didn't realize this was an issue until this github issue was made, but we are currently working on a fix for it. "
And if yes when do you think you will have a release solving this?
Many thanks

**Comment by yangpei1010110** [comment]
you need delete Microsoft.ML.OnnxRuntime package and using Microsoft.ML.OnnxRuntime.Gpu package.

**Comment by yangpei1010110** [comment]
and input gpu index
`MLContext mlContext = new MLContext();
            var transformer = mlContext.Transforms
                .ResizeImages(outputColumnName: "images", imageWidth: modelWidth, imageHeight: modelHeight, inputColumnName: "Image",
                    resizing: ImageResizingEstimator.ResizingKind.IsoPad)
                .Append(mlContext.Transforms.ExtractPixels("images", scaleImage: 1f / 255f, interleavePixelColors: false))
                .Append(mlContext.Transforms.ApplyOnnxModel(
                    gpuDeviceId: 0, // gpu index
                    shapeDictionary: new Dictionary<string, int[]>()
                    {
                        {"images", new[] {1, 3, modelWidth, modelHeight}}
                    },
                    outputColumnName: "output",
                    inputColumnName: "images",
                    modelFile: modelFilePath))
                .Fit(mlContext.Data.LoadFromEnumerable(new List<InputData>()));`

**Comment by yangpei1010110** [comment]
```
MLContext mlContext = new MLContext();
            var transformer = mlContext.Transforms
                .ResizeImages(outputColumnName: "images", imageWidth: modelWidth, imageHeight: modelHeight, inputColumnName: "Image",
                    resizing: ImageResizingEstimator.ResizingKind.IsoPad)
                .Append(mlContext.Transforms.ExtractPixels("images", scaleImage: 1f / 255f, interleavePixelColors: false))
                .Append(mlContext.Transforms.ApplyOnnxModel(
                    gpuDeviceId: 0, // gpu index
                    shapeDictionary: new Dictionary<string, int[]>()
                    {
                        {"images", new[] {1, 3, modelWidth, modelHeight}}
                    },
                    outputColumnName: "output",
                    inputColumnName: "images",
                    modelFile: modelFilePath))
                .Fit(mlContext.Data.LoadFromEnumerable(new List<InputData>()));
```

**Comment by michaelgsharp** [comment]
@sportbilly21 the fix for the onnx runtime code is now in a PR https://github.com/dotnet/machinelearning/pull/6143/files. So it should be pretty soon that code is checked into main.

For tensorflow I'm not sure. Did you add the tensorflow gpu package? and do you have cuda 10.1 installed?

**Comment by sportbilly21** [comment]
Hi
@michaelgsharp  many thanks for the update. Yes I have installed all the correct Cuda and CuDnn packages.
Regarding the fix in the ML.Net how soon think you will have a new Nuget package released?

**Comment by michaelgsharp** [comment]
@sportbilly21 I'm not sure when the next servicing release is. I'll figure that out and let you know.

Would you be able to test it with the latest nightly release that we have and see if it helps with the ONNX performance? If it doesn't that means there is more we need to look into. If it does than the GPU fix is enough. In the latest nightly there is now a flag in the MLContext you an use to set the GPU device ID.

Are you trying to predict multiple images? Or are you trying to do only 1 image and even that is pretty slow?

**Comment by michaelgsharp** [comment]
So the flag in the MLContext is the same flag as in ApplyOnnxModel. It just you can set it directly in the MLContext.

As far as WindowsML is concerned, do you know if the windows ML onnx runtime api is the same as the normal onnx runtime api? If it is you could probably just do a drop in replacement for it. I know we have talked about WinML before, but I'm not sure what our current status on it is.

@luisquintanilla for visibility about WinML.
